torchrl/envs/libs/gym.py

```diff
@@ -216,16 +226,28 @@ def _gym_to_torchrl_spec_transform(
             spec.n, device=device, dtype=numpy_to_torch_dtype_dict[spec.dtype]
         )
     elif isinstance(spec, gym.spaces.multi_discrete.MultiDiscrete):
-        dtype = (
-            numpy_to_torch_dtype_dict[spec.dtype]
-            if categorical_action_encoding
-            else torch.long
-        )
-        return (
-            MultiDiscreteTensorSpec(spec.nvec, device=device, dtype=dtype)
-            if categorical_action_encoding
-            else MultiOneHotDiscreteTensorSpec(spec.nvec, device=device, dtype=dtype)
+        # dtype = (
+        #     numpy_to_torch_dtype_dict[spec.dtype]
+        #     if categorical_action_encoding
+        #     else torch.long
+        # )
+        return torch.stack(
+            [
+                _gym_to_torchrl_spec_transform(
+                    spec[i],
+                    device=device,
+                    categorical_action_encoding=categorical_action_encoding,
+                    remap_state_to_observation=remap_state_to_observation,
+                )
+                for i in range(len(spec.nvec))
+            ],
+            0,
         )
+        # return (
+        #     MultiDiscreteTensorSpec(spec.nvec, device=device, dtype=dtype)
+        #     if categorical_action_encoding
+        #     else MultiOneHotDiscreteTensorSpec(spec.nvec, device=device, dtype=dtype)
+        # )
```

Comment:
wait why would it be removed? this is actually very needed and important

Reply:
what's the difference between that and a stack of discrete specs?

Reply:
Thing is that a MultiDiscrete(2, 2) has a batch-size of 2, but that is being translated to a MultiOneHot of size (4,), which is batch-size incompatible

Reply:
there are envs where you need to select two actions at the same time (let's say one for movement and one for comms)

those 2 actions might have different spaces. for example in vmas movement is of space 4 and comms can be anything (for example 10). why would we stack this?

if the problem here is that they are using multidiscrete to indicate batched envs i think they overloaded multidiscrete then, because that should just be a batched discrete

Reply:
Here's a comprehensive example after my latest push

```python
from gymnasium.spaces import MultiDiscrete

from torchrl.envs.libs.gym import _gym_to_torchrl_spec_transform

disc = MultiDiscrete([[2, 3], [2, 3]])

print("multi-discrete, gym", disc)
print("nvec shape", disc.nvec.shape)
spec = _gym_to_torchrl_spec_transform(disc, categorical_action_encoding=True)
print("spec, discrete", spec)
print("sample", spec.rand())
print("spec shape", spec.shape)
spec = _gym_to_torchrl_spec_transform(disc, categorical_action_encoding=False)
print("spec, one-hot", spec)
print("sample", spec.rand())
print("spec shape", spec.shape)

disc = MultiDiscrete([[2, 3], [4, 5]])

print("multi-discrete, gym", disc)
print("nvec shape", disc.nvec.shape)
spec = _gym_to_torchrl_spec_transform(disc, categorical_action_encoding=True)
print("spec, discrete", spec)
print("sample", spec.rand())
print("spec shape", spec.shape)
spec = _gym_to_torchrl_spec_transform(disc, categorical_action_encoding=False)  # breaks
print("spec, one-hot", spec)
print("sample", spec.rand())
print("spec shape", spec.shape)
```

the print looks like this
```
multi-discrete, gym MultiDiscrete([[2 3]
 [2 3]])
nvec shape (2, 2)
spec, discrete MultiDiscreteTensorSpec(
    shape=torch.Size([2, 2]),
    space=BoxList(boxes=[BoxList(boxes=[DiscreteBox(n=2), DiscreteBox(n=2)]), BoxList(boxes=[DiscreteBox(n=3), DiscreteBox(n=3)])]),
    device=cpu,
    dtype=torch.int64,
    domain=discrete)
sample tensor([[0, 0],
        [1, 2]])
spec shape torch.Size([2, 2])
spec, one-hot MultiOneHotDiscreteTensorSpec(
    shape=torch.Size([2, 5]),
    space=BoxList(boxes=[DiscreteBox(n=2), DiscreteBox(n=3)]),
    device=cpu,
    dtype=torch.int64,
    domain=discrete)
sample tensor([[0, 1, 1, 0, 0],
        [0, 1, 0, 1, 0]])
spec shape torch.Size([2, 5])
multi-discrete, gym MultiDiscrete([[2 3]
 [4 5]])
nvec shape (2, 2)
spec, discrete LazyStackedMultiDiscreteTensorSpec(
    shape=torch.Size([2, 2]), device=cpu, dtype=torch.float32, domain=discrete)
sample tensor([[1, 0],
        [0, 0]])
spec shape torch.Size([2, 2])
spec, one-hot LazyStackedMultiOneHotDiscreteTensorSpec(
    shape=torch.Size([2, -1]), device=cpu, dtype=torch.float32, domain=discrete)
```

@matteobettini what should we do with the example that breaks?

Reply:
To resolve this issue:
- Can you provide an example of where you expect that things will break and/or the user experience will be negatively impacted. Here is my 2 cents:
  - MultiOneHot is a giant hack. Per se, you work with a tensor that is strongly tied to its spec. Without the spec, you cannot know what this tensor represents: the only instance I see where you know how to split it is all cases like `[False, False, True, True, False]`. So because _if_ you're using MultiOneHot you _must_ use the spec wherever you use the tensor, _then_ when you use the simplified version (with OneHot + batch-size) you will also have the spec at hand (because I presume that your code base handles the most complex use case so the simpler one falls into the same requisites). Therefore, anything you do with your tensor can probably be represented with methods from the spec (e.g. `spec.encode`, `spec.to_numpy` etc) and you very rarely interact directly with the tensor. I do not see anywhere where a control flow would be added like `if isinstance(spec, MultiOneHot): ... ; else ...`. 
  - Along the same line, it isn't easy to work with MultiOneHot and I'm not sure I've ever seen such a data structure anywhere else than torchrl. So from a UX perspective restricting its usage to edge cases isn't a great price to pay.
  - It is important also to acknowledge what percentage of users (roughly) will be impacted by which decision. MultiOneHot is to be used by the following subset of users: those who work with more than one agent/env + have action space with heterogeneous action + want to use one-hot encoding (and not discrete). Counting that each of these is roughly 50% of users, deciding that the behaviour changes when agents are homogeneous or heterogeneous impacts at the most 0.5 ** 3 = 0.125 of our user base (and this is assuming that there actually is a negative impact for these people, but I haven't really understood what the negative impact is yet).
  - I see the point that a different config will create different classes depending on homogeneity. On the other hand, this is the case in many paradigm (e.g. pytorch dataloaders are from different classes if `num_workers=0` or `num_workers=2` and people are usually not complaining about it much because the way they work with the dataloader isn't impacted).

The most important thing is what solution we can provide to this problem:
- What is the way forward to you? Here are the options I see
  - Leave as it. This means that with Gymnasium, vectorized envs will have a `MultiOneHot(num_envs * num_action)` spec which we will have a hard time fitting in the env.
  - Have a distinct behaviour for gymnasium, gym and other libs like VMAS. This will be a high price to pay in term of UX: we'll need to document that with gym one should expect the conversion to be like X->Y, with VMAS X->Z and with gymnasium X->W. Therefore, we gain the consistency within a domain (eg, MARL) but loose it across libs, which is one feature we want to promote about torchrl I think.
  - Use the proposed with the following caveats:
    - If a MultiOneHot is created, the user is informed once in the execution that this is the spec they will get. We explicitly say why they should not use one-hot (performance is poorer, performance degrates as the action space increases and stacking heterogeneous creates a structure that is hard to work with)
    - In a not so distant future we switch the default to discrete and make batched discrete fall back onto multi-discrete (I think it's already the case).

Reply:
I think the current proposed solution is the best in order to be able to handle gymnasium vectorised environments that are using `MultiDiscrete` spaces to represent the vectorisation dimension.
We can take this route and take some precautions as discussed, for example:

- Document the spec translation mechanism 
- Default to categorical action encoding

### Example
My example about a potential surprising behaviour is the following (it is in the single-agent domain, so it is not related to MARL):
```python
spec = gym.spec.MultiDiscrete(n1, n2) # Where n1 is for example the number of movement actions and n2 is the number of communication actions
```
Now, when this is translated in torchrl, with action encoding set to one-hot, it will give the following specs:
```python
if n1 == n2:
    OneHotTensorSpec(shape=(2, n1)) 
else:
    MultiOneHotTensorSpec(shape=(n1+n2))
```
So simplifying the spec to a OneHot in case the action sizes are the same.

That said, I am on board with the current solution for gymnasium spec translation as in the above example we do not have the tools to understand if we are seeing a single environment with multiple discrete actions or a vectorized environment with a single discrete action.

Reply:
Cool I'll check on some edge cases and ping you with these!